critwall/src/main/java/org/opencadc/critwall/FileSyncJob.java

```diff
@@ -0,0 +1,256 @@
+
+            // sync succeeded - update storage location
+            if (storageMeta != null) {
+                log.debug(storageMeta.getContentChecksum());
+                log.debug(storageMeta.contentLastModified);
+                log.debug(storageMeta.getContentLength());
+
+                // Update curArtifact with new storage location
+                Artifact curArtifact = artifactDAO.get(artifactID);
+                curArtifact.storageLocation = storageMeta.getStorageLocation();
+                artifactDAO.put(curArtifact, true);
+                log.debug("updated artifact storage location");
+                break;
+            }
+            retryCount++;
+        }
+    }
+
+    private List<URL> getdownloadURLs() throws IOException, InterruptedException,
+        ResourceAlreadyExistsException, ResourceNotFoundException,
+        TransientException, TransferParsingException {
+
+        RegistryClient regClient = new RegistryClient();
+        log.debug("resource id: " + this.resourceID);
+        URL transferURL = regClient.getServiceURL(this.resourceID, Standards.VOSPACE_SYNC_21, AuthMethod.CERT);
+        log.debug("certURL: " + transferURL);
+
+        // Ask for all protocols available back, and the server will
+        // give you URLs for the ones it allows.
+        List<Protocol> protocolList = new ArrayList<Protocol>();
+        Protocol httpsCert = new Protocol(VOS.PROTOCOL_HTTPS_GET);
+
+        log.debug("protocols: " + httpsCert);
+        protocolList.add(httpsCert);
+
+        Transfer transfer = new Transfer(artifactID, Direction.pullFromVoSpace, protocolList);
+        transfer.version = VOS.VOSPACE_21;
+
+        TransferWriter writer = new TransferWriter();
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        writer.write(transfer, out);
+        FileContent content = new FileContent(out.toByteArray(), "text/xml");
+        log.debug("xml file content to be posted: " + transfer);
+
+        log.debug("artifact path: " + artifactID.getPath());
+        HttpPost post = new HttpPost(transferURL, content, true);
+        post.prepare();
+        log.debug("post prepare done");
+
+        TransferReader reader = new TransferReader();
+        Transfer t = reader.read(post.getInputStream(), null);
+        List<String> urlStrList = t.getAllEndpoints();
+        log.debug("endpoints returned: " + urlStrList);
+
+        List<URL> urlList = new ArrayList<URL>();
+
+        // Create URL list to return
+        for (int i = 0; i < urlStrList.size(); i++) {
+            try {
+                urlList.add(new URL(urlStrList.get(i)));
+            } catch (MalformedURLException mue) {
+                log.info("Malformed URL returned from transfer negotiation: " + urlStrList.get(i) + " skipping... ");
+            }
+        }
+
+        return urlList;
+    }
+
+    private StorageMetadata syncArtifact(Iterator<URL> urlIterator) {
+        StorageMetadata storageMeta = null;
+
```

Comment:
This looks pretty good now and is easy to grok. I would still improve it further with:

line ~228 : immediately return the StorageMetadata object rather than `break`
line ~230 : in the first catch, use multi-catch for all the exceptions where you would just continue trying (TransientException, ReadException,  IOException for now), don't need `continue`
line ~235 : you should call urlIterator.remove() here so that URL is gone if we ever add any kind of pause/resume or longer retry delay
line ~239 : I would just have this method let the fatal errors propagate up (StorageEngageException, WriteException, InterruptedException)

With that, you won't need the return at the end and can never return null.

Reply:
IOException will catch the WriteException (which would turn it into a transient instead of a failure situation), if I'm reading the error my IDE reports correctly?

Returning null is the only thing that would trigger the outer while loop to retry the entire list. If no storageMeta is returned, then none of the URLs resulted in a successful get. The catastrophic fails already quit, but if there are transient errors for each url, there has to be a case where a null can fall through and trigger a retry on the whole list. (Or am I misunderstanding the retry idea?)

Reply:
changes made, but the WriteException is not handled symmetrically with the rest of the exceptions because of the IOException swallowing it unless it's trapped explicitly.

Reply:
Write Exception is now thrown as a fatal error.